libass/ass_fontselect.c

```diff
@@ -608,9 +606,12 @@ void ass_font_provider_free(ASS_FontProvider *provider)
 static bool check_postscript(ASS_FontInfo *fi)
 {
     ASS_FontProvider *provider = fi->provider;
-    assert(provider && provider->funcs.check_postscript);
+    assert(provider);
 
-    return provider->funcs.check_postscript(fi->priv);
+    if (provider->funcs.check_postscript)
+        return provider->funcs.check_postscript(fi->priv);
+    else
+        return fi->is_postscript;
 }
```

Comment:
Since I didn't spot any `if (funcs.check_postscript)` guards elsewhere, do I understand correctly `fi->is_postscript` is already computed regardless of whether the font provider has `check_postscript?
If so, what’s the benefit of a provider having `check_postscript`? And do the performance considerations for DirectWrite’s  eager loading also apply to Fontconfig’s still existing eager loading on startup?

Reply:
`fi->is_postscript` always exists, but it only has a meaningful value for font providers that invoke `get_font_info` one way or another during the addition of the font. Currently, that’s Core Text and the memory font provider. To declare this fact, these providers don’t define a `check_postscript` function (which would be useless after `is_postscript` is already computed).

Potentially, the value returned by `check_postscript` could be cached in `is_postscript` for the other providers. But this would require `is_postscript` to be something more complex than a simple `bool`: perhaps a tri-value `enum` or a pair of Boolean bit-fields. My first attempt actually was something like that (minus the caching), until I realized I could simplify it to this.

> what’s the benefit of a provider having `check_postscript`?

It lets the provider avoid expensive operations to figure out this value until the font actually matches some requested name.

For example, this presumably still applies (taken in isolation) to the DirectWrite provider just as it did back then: when enumerating fonts, DirectWrite seems to operate on a fast cache, but that cache doesn’t include the font format information. To get that, `check_postscript` must materialize the cache entry to the actual font and query the font for its format, but this is extra slow: per the original commit message, this step alone was apparently costing seconds for a font library that must’ve been small enough that it was otherwise loading reasonably quickly. Of course, when we actually want to select and to start using a font, we must materialize it anyway, but this lets us avoid materializing DirectWrite fonts that we are definitely not interested in.

Now, with the DirectWrite font provider now adding fonts lazily, this performance gain may have become negligible by Amdahl’s law. In fact, when we lazy-add fonts, we look them up by name, so in the ideal case, every lazily-added font will be a name match and cause `check_postscript` to run and/or the font to be selected and materialized for use. But technically, it’s possible for the font provider to return more fonts due to matching some name that we choose to ignore; although then again, I don’t remember if this is possible for the DirectWrite provider in particular.

Being unsure, I chose to keep this PR as small & safe as possible.

> And do the performance considerations for DirectWrite’s eager loading also apply to Fontconfig’s still existing eager loading on startup?

Nobody’s ever reported any performance issues with the Fontconfig system font provider.

Reply:
Thanks for the detailed explanation!